### PR TITLE
Add extrema_filter

### PR DESCRIPTION
**Pull request by codles**
Added function and tests for 2d and 4d case

### CODE REVIEW

**Comment by codles** [comment]
Initial add following [METADATA](https://github.com/JuliaLang/METADATA.jl/pull/4036) discussion.

File: src/algorithms.jl [hunk 1]
```diff
@@ -1473,6 +1473,171 @@ extr(order::Ordering, x::Color, y::Color) = extr(order, convert(RGB, x), convert
...
+                # Create index for full array (fa) length
+                @nexprs $(N)   j->(fa_{j} = 1:size(maxval_temp)[j])
+                @nexprs $(N-1) j->(fa_{j} = i_j)
+
+                # Create index for short array (sa) length
+                @nexprs $(N)   j->(sa_{j} = 1:size(maxval_temp)[j] - window[dim] + 1)
+                @nexprs $(N-1) j->(sa_{j} = i_j)
+
+                # Filter the last dimension
+                (@nref $N minval_temp sa) = min_filter(vec( @nref $N minval_temp fa), window[dim])
```
Comment (timholy) [inline src/algorithms.jl]: This doesn't seem to be defined (likewise `max_filter`)

File: src/algorithms.jl [hunk 2]
```diff
@@ -1473,6 +1473,171 @@ extr(order::Ordering, x::Color, y::Color) = extr(order, convert(RGB, x), convert
 extr(order::Ordering, x::Color, y::Color, z::Color) = extr(order, convert(RGB, x), convert(RGB, y), convert(RGB, z))
 
 
+
+# Min max filter
+
+type Wedge
+    buffer::AbstractArray
```
Comment (timholy) [inline src/algorithms.jl]: For good performance, you're going to want to declare this to be a leaf type. If you need to handle arbitrary array types, then something like
```jl
type Wedge{A<:AbstractArray}
    buffer::A
    size::Int
    ...
end
```
will give you much better performance.

Comment (timholy) [inline src/algorithms.jl]: (See http://docs.julialang.org/en/stable/manual/faq/#how-do-abstract-or-ambiguous-fields-in-types-interact-with-the-compiler)

File: src/algorithms.jl [hunk 3]
```diff
@@ -1473,6 +1473,171 @@ extr(order::Ordering, x::Color, y::Color) = extr(order, convert(RGB, x), convert
...
+
+    # Initialise the internal wedges
+    # U[1], L[1] are the location of the global maximum and minimum
+    # U[2], L[2] are the maximum and minimum over (U1, inf)
+    L = Wedge(zeros(Int,1,window+1), window+1, 0, 1, 0, 0)          # Min
+    U = Wedge(zeros(Int,1,window+1), window+1, 0, 1, 0, 0)
+
+    for i = 2:n
+        if i > window
+            if ~wedgeisempty(U)
```
Comment (timholy) [inline src/algorithms.jl]: ! rather than ~? (Matlab-ism?)

File: src/algorithms.jl [hunk 4]
```diff
@@ -1473,6 +1473,171 @@ extr(order::Ordering, x::Color, y::Color) = extr(order, convert(RGB, x), convert
...
+
+    return minval, maxval
+end
+
+function wedgeisempty(X::Wedge)
+    X.n <= 0
+end
+
+function pushback(X::Wedge, v)
+    X.last = mod(X.last, X.size) + 1
```
Comment (timholy) [inline src/algorithms.jl]: AFAICT, this is just a circular buffer and loops 1,2,3,...,size,1,2,3,...size,.... It turns out that `mod` is slow (it relies on integer division, which is one of the slowest operations on a CPU), so you could probably speed this up with
```jl
X.last += 1
if X.last > X.size
    X.last = 1
end
```
or something.

File: src/algorithms.jl [hunk 5]
```diff
@@ -1473,6 +1473,171 @@ extr(order::Ordering, x::Color, y::Color) = extr(order, convert(RGB, x), convert
...
+    X.buffer[X.last] = v
+    X.n = X.n+1
+    X.mxn = maximum([X.mxn, X.n])
```
Comment (timholy) [inline src/algorithms.jl]: `max(X.mxn, X.n)` will be orders-of-magnitude faster.

File: src/algorithms.jl [hunk 6]
```diff
@@ -1473,6 +1473,171 @@ extr(order::Ordering, x::Color, y::Color) = extr(order, convert(RGB, x), convert
...
+    end
+
+    return minval, maxval
+end
+
+function wedgeisempty(X::Wedge)
+    X.n <= 0
+end
+
+function pushback(X::Wedge, v)
```
Comment (timholy) [inline src/algorithms.jl]: Maybe call this `pushback!`? Functions which mutate their inputs conventionally end in `!`.

**Comment by codles** [comment]
Thanks @timholy. Ill make your suggested changes. And a few changes of my own.
Should I build on this pull request? Or make the changes and open a new pull request?

File: src/algorithms.jl [hunk 7]
```diff
@@ -1473,6 +1473,171 @@ extr(order::Ordering, x::Color, y::Color) = extr(order, convert(RGB, x), convert
...
+
+                # Create index for full array (fa) length
+                @nexprs $(N)   j->(fa_{j} = 1:size(maxval_temp)[j])
+                @nexprs $(N-1) j->(fa_{j} = i_j)
+
+                # Create index for short array (sa) length
+                @nexprs $(N)   j->(sa_{j} = 1:size(maxval_temp)[j] - window[dim] + 1)
+                @nexprs $(N-1) j->(sa_{j} = i_j)
+
+                # Filter the last dimension
```
Comment (timholy) [inline src/algorithms.jl]: It seems that you're actually filtering on the first dimension? Julia arrays are column-major.

**Comment by timholy** [comment]
You can just build on this one. When you push new commits to `codles:master`, this will be automatically updated.

I don't get an email each time you push a commit, so when you're ready for me to check this again just post a comment here.

File: src/algorithms.jl (hunk 2 above)
Comment (codles) [inline src/algorithms.jl]: Thanks

File: src/algorithms.jl (hunk 3 above)
Comment (codles) [inline src/algorithms.jl]: Thanks. Old habits die hard

File: src/algorithms.jl (hunk 5 above)
Comment (codles) [inline src/algorithms.jl]: I didn't know this. Thanks

File: src/algorithms.jl (hunk 6 above)
Comment (codles) [inline src/algorithms.jl]: Noted, thanks

File: src/algorithms.jl (hunk 4 above)
Comment (codles) [inline src/algorithms.jl]: Again I was unaware of this. Thanks

**Comment by codles** [comment]
@timholy I think I have addressed your concerns above and added tests for all lines of code. Please let me know of any additional changes you want made.

**Comment by timholy** [comment]
Yay, thanks!

**Pull request merged**